### PR TITLE
Keystone plugin

### PR DESCRIPTION
**Pull request by ghost**
## Overview

Creating a Keystone Plugin for usage again `provenance-keystone`. Since repo is private to Figure, re-implementing the api client via Retrofit. 

Note: will fail build / won't check this in until related `scope-sdk` changes are published.

### CODE REVIEW

File: lib/src/main/kotlin/io/provenance/plugins/keystone/model/SignatureRequest.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package io.provenance.plugins.keystone.model
+
+data class SignatureRequest(
+    val data: ByteArray,
+    val signingType: SigningType,
+)
+
+enum class SigningType {
```
Comment (piercetrey-figure) [inline lib/src/main/kotlin/io/provenance/plugins/keystone/model/SignatureRequest.kt:8]: what's the difference between these two signing types in Keystone?

File: lib/src/main/kotlin/io/provenance/plugins/keystone/KeystonePlugin.kt [hunk 2]
```diff
@@ -2,9 +2,23 @@ package io.provenance.plugins.keystone
...
 import io.provenance.core.Plugin
 import io.provenance.entity.KeyEntity
+import io.provenance.entity.KeyType
+import io.provenance.entity.keystone.KeystoneKeyEntity
+import io.provenance.scope.encryption.ecies.ECUtils
+import io.provenance.scope.encryption.model.ApiKeyRef
 
 class KeystonePlugin: Plugin<KeyStoneConfig> {
     override fun fetch(entity: String, config: KeyStoneConfig): KeyEntity {
-        TODO("Not yet implemented")
```
Comment (piercetrey-figure) [inline lib/src/main/kotlin/io/provenance/plugins/keystone/KeystonePlugin.kt:8]: 🎉 TADA 🎉

File: lib/src/main/kotlin/io/provenance/plugins/keystone/KeystoneClient.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    .configureProvenance()
+                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false),
+            )
+        )
+        .build()
+        .create(KeystoneApi::class.java)
+
+    override fun sign(data: ByteArray): ByteArray {
+        val request = SignatureRequest(data, SigningType.PB)
+        val response = client.sign(apikey, entity,0, request)
```
Comment (dkneisly-figure) [inline lib/src/main/kotlin/io/provenance/plugins/keystone/KeystoneClient.kt]:
```suggestion
        val response = client.sign(apikey, entity, 0, request)
```

File: lib/src/main/kotlin/io/provenance/plugins/keystone/KeystoneClient.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        if (!response.isSuccessful) {
+            throw IllegalStateException("Failed to sign with error ${response.errorBody()}")
+        }
+
+        return response.body()?.signatureBytes!!
+    }
+
+    override fun secretKey(ephemeralPublicKey: PublicKey): ByteArray {
+        val request = AgreeKeyRequest(entity, ECUtils.convertPublicKeyToBytes(ephemeralPublicKey),0)
```
Comment (dkneisly-figure) [inline lib/src/main/kotlin/io/provenance/plugins/keystone/KeystoneClient.kt]:
```suggestion
        val request = AgreeKeyRequest(entity, ECUtils.convertPublicKeyToBytes(ephemeralPublicKey), 0)
```

File: lib/src/main/kotlin/io/provenance/plugins/keystone/model/SignatureRequest.kt (hunk 1 above)
Comment (ghost) [reply lib/src/main/kotlin/io/provenance/plugins/keystone/model/SignatureRequest.kt:8]: note entirely sure, but looks like p8e is unimplemented https://github.com/FigureTechnologies/provenance-keystone/blob/ee610b7fec1ed51ad2472ed568e8ebd5e89c6063/service/src/main/kotlin/io/provenance/keystone/service/KeystoneService.kt#L177

File: lib/src/main/kotlin/io/provenance/plugins/keystone/KeystoneClient.kt (hunk 3 above)
Comment (piercetrey-figure) [reply lib/src/main/kotlin/io/provenance/plugins/keystone/KeystoneClient.kt]: ol' @dkneisly-figure the linter we call him! 😄

**Review by piercetrey-figure — DISMISSED** [review]

**Review by dkneisly-figure — APPROVED** [review]

**Pull request merged**